DOC: Add documentation for cudf.pandas in the Developer Guide (#15889)
rapidsai/cudf: docs/cudf/source/developer_guide/cudf_pandas.md

# cudf.pandas
The use of the cuDF pandas accelerator mode (`cudf.pandas`) is explained [here](../cudf_pandas/index.rst).
The purpose of this document is to explain how the fast-slow proxy mechanism works and document the internal environment variables used to debug `cudf.pandas`.

## fast-slow proxy mechanism:

> `cudf.pandas` works by wrapping each Pandas type and its corresponding cuDF type in new proxy types (aka fast-slow proxy types).

Author: Thanks for this. Just to be clear, would numpy and cupy be another example?

> `cudf.pandas` works by wrapping each Pandas type and its corresponding cuDF type in new proxy types (aka fast-slow proxy types).

Reviewer: For example, yes

Intermediate types are the types of the results of operations invoked on final types.
For example, `xpd.DataFrameGroupBy` is an intermediate type that will be created during a groupby operation on the final type `xpd.DataFrame`.

##### Attributes and Callable Proxy Types:
Final proxy types are typically classes or modules, both of which have attributes.
Classes also have methods.
These attributes and methods must be wrapped as well to support the fast-slow proxy scheme.

#### Creating New Proxy Types:
`_FinalProxy` and `_IntermediateProxy` types are created using the functions `make_final_proxy_type` and `make_intermediate_proxy` type, respectively. Creating a new final type looks like this.

```python
DataFrame = make_final_proxy_type(
    "DataFrame",
    cudf.DataFrame,
    pd.DataFrame,
    fast_to_slow=lambda fast: fast.to_pandas(),
    slow_to_fast=cudf.from_pandas,
)
```

### The Fallback Mechanism
The `_fast_slow_function_call` is the all-important function that we use to call operations the fast way (using cuDF) and if that fails, the slow way (using Pandas).

> The is also known as the fallback mechanism.

Reviewer: I would perhaps write something like:

> Proxied calls are implemented with fallback via `_fast_slow_function_call`. This implements the mechanism by which we attempt operations the fast way (using cuDF) and then fall back to the slow way (using pandas) on failure.

The function looks like this:
```python
def _fast_slow_function_call(func: Callable, *args, **kwargs):
    try:
        ...
        fast_args, fast_kwargs = _fast_arg(args), _fast_arg(kwargs)
        result = func(*fast_args, **fast_kwargs)
        ...
    except Exception:
        ...
        slow_args, slow_kwargs = _slow_arg(args), _slow_arg(kwargs)
        result = func(*slow_args, **slow_kwargs)
        ...
    return _maybe_wrap_result(result, func, *args, **kwargs), fast
```

> As you can see the function attempts to call `func` the fast way using cuDF and if any `Exception` occurs, it calls the function using Pandas.

Reviewer: ```suggestion
As we can see the function attempts to call `func` the fast way using cuDF and if any `Exception` occurs, it calls the function using Pandas.
```

Although I know I might be the minority in preferring the active "we" form in documentation.

> In essence, this `try-except` is what allows `cudf.pandas` to support 100% of the Pandas API.

Author: I added a permalink to the function

At the end, the function wraps the result from either path in a fast-slow proxy object, if necessary.

#### Converting Proxy Objects
Note that before the `func` is called, the proxy object and its attributes need to be converted to either their cuDF or Pandas implementations.
This conversion is handled in the function `_transform_arg` which both `_fast_arg` and `_slow_arg` call.

`_transform_arg` is a recuirsive function that will call itself depending on the type or argument passed to it (eg. `_transform_arg` is called for each element in a list of arguments).

### Using Metaclasses

> `cudf.pandas` uses a [metaclass](https://docs.python.org/3/glossary.html#term-metaclass) called (`_FastSlowProxyMeta`) to dynamically find class attributes and classmethods of fast-slow proxy types.

Reviewer 2: Dynamically is a bit of a loaded term here, let's avoid it. It's not obvious exactly what dynamic means in this context (for example, after https://github.com/rapidsai/cudf/pull/14534/ we no longer wait until an attribute is queried to do the lookup, it happens on type creation).
```suggestion
`cudf.pandas` uses a [metaclass](https://docs.python.org/3/glossary.html#term-metaclass) called (`_FastSlowProxyMeta`) to find class attributes and classmethods of fast-slow proxy types.
```

For example, in the snippet below, the `xpd.Series` type is an instance of `_FastSlowProxyMeta`.
Therefore we can access the property `_fsproxy_fast` defined in the metaclass.
```python
import cudf.pandas
cudf.pandas.install()
import pandas as xpd

print(xpd.Series._fsproxy_fast) # output is cudf.core.series.Series
```

## debugging `cudf.pandas`

> Several environment variables are available for debugging purposes.

Reviewer: *question*: What are the other ones?

Author: The next one (or possibly ones) I'm adding is for debugging different types of fallback (eg. OOM errors, NotImplementedErrors). I'm working on it #15910.

Reviewer: Ah thanks.

# cudf.pandas

> The use of the cuDF pandas accelerator mode (`cudf.pandas`) is explained [here](https://docs.rapids.ai/api/cudf/stable/cudf_pandas/). The purpose of this document is to explain how the fast-slow proxy mechanism works and document the internal environment variables used to debug `cudf.pandas`.

Reviewer 2: Instead of linking the URL, we want to link to the page within the doc. We use MyST to generate our documentation, so I would take a look at [the cross-linking guide](https://mystmd.org/guide/cross-references) for how to do that here.

## fast-slow proxy mechanism

> `cudf.pandas` works by wrapping each Pandas type (and its corresponding cuDF type) in new proxy types (aka fast-slow proxy types). Because the proxy types wrap both the fast and slow implementations of the original type, we can ensure that computations are first done on the fast version of the proxy type, and if that fails, the slow version of the proxy type.

Reviewer 2: ```suggestion
`cudf.pandas` works by wrapping each Pandas type and its corresponding cuDF type in new proxy types (aka fast-slow proxy types). The purpose of the proxy types is to attempt computations on the fast (cuDF) object first, and then fall back to running on the slow type if the fast version fails.
```

### Types
#### Wrapped Types and Proxy Types
The "wrapped" types/classes are the Pandas and cuDF specific types that have been wrapped into proxy types. Wrapped objects and proxy objects are instances of wrapped types and proxy types, respectively. In the snippet below `s1` and `s2` are wrapped objects and `s3` is a fast-slow proxy object. Also note that the module `xpd` is a wrapped module and contains cuDF and Pandas modules as attributes.
  ```python
  import cudf.pandas
  cudf.pandas.install()
  import pandas as xpd

  cudf = xpd._fsproxy_fast
  pd = xpd._fsproxy_slow

  s1 = cudf.Series([1,2])
  s2 = pd.Series([1,2])
  s3 = xpd.Series([1,2])
  ```

> 

Reviewer 2: Probably worth adding a small note that users are never expected to have to access the wrapped types directly, and that this code is just for demonstration.
```suggestion

Note that users should never have to interact with the wrapped objects directly in this way. This code is purely for demonstrative purposes.

```

#### The Different Kinds of Proxy Types
In `cudf.pandas`, proxy types come in multiple kinds: fast-slow proxy types, callable types, and fast-slow attribute types.

Fast-slow proxy types come in two flavors: final types and intermediate types. Final types are types for which known operations exist for converting an object of "fast" type to "slow" and vice-versa. For example, `cudf.DataFrame` can be converted to Pandas using the method `to_pandas` and `pd.DataFrame` can be converted to cuDF using the function `cudf.from_pandas`.

Intermediate types are the types of the results of operations invoked on final types. For example, `DataFrameGroupBy` is a type that will be created during a groupby operation.

Callable types are ...

> Fast-Slow attribute types ...

Reviewer 2: I would suggest framing this a little bit differently. Rather than discussing the three different types that you have, I would start with describing the difference between final and intermediate proxy types. Then, talk about the fact that final proxy types are typically classes or modules, both of which have that have attributes (classes also have methods and properties). These must be wrapped as well, which is where the fast-slow attributes and callable proxies come in (we use a callable proxy both for class methods and free functions, the latter of which are essentially attributes of a module).

#### Creating New Proxy Types
`_FinalProxy` and `_IntermediateProxy` types are created using the functions `make_final_proxy_type` and `make_intermediate_proxy` type, respectively. Creating a new final type looks like this.

```python
DataFrame = make_final_proxy_type(
    "DataFrame",
    cudf.DataFrame,
    pd.DataFrame,
    fast_to_slow=lambda fast: fast.to_pandas(),
    slow_to_fast=cudf.from_pandas,
)
```

### The Fallback Mechanism
The `_fast_slow_function_call` is the all-important function that we use to call operations the fast way (using cuDF) and if that fails, the slow way (using Pandas). The is also known as the fallback mechanism. The function looks like this:
```python
def _fast_slow_function_call(func: Callable, *args, **kwargs):
    try:
        ...
        fast_args, fast_kwargs = _fast_arg(args), _fast_arg(kwargs)
        result = func(*fast_args, **fast_kwargs)
        ...
    except Exception:
        ...
        slow_args, slow_kwargs = _slow_arg(args), _slow_arg(kwargs)
        result = func(*slow_args, **slow_kwargs)
        ...
```

> As you can see the function attempts to call `func` the fast way using cuDF and if an any `Exception` occurs, it calls the function using Pandas. In essence, this `try-except` is what allows `cudf.pandas` to support 100% of the Pandas API.

Reviewer 2: Another useful note here is that everywhere that `_fast_slow_function_call` is invoked, the current object is also passed to the function. A good example is in the `__init__` method assigned to each final proxy type. The reason that this is important is that this is where the magic of converting proxy objects happens: the `_fast_arg` and `_slow_arg` calls above aren't just converting arguments to a method call, they actually convert the object whose method is being invoked as well. That's easy to miss on first read.

Author: I wasn't clear on how the type conversion worked. Thanks for the note!

### Using Metaclasses

> `cudf.pandas` uses a [metalass](https://docs.python.org/3/glossary.html#term-metaclass) called (`_FastSlowProxyMeta`) to dynamically find class attributes and classmethods of fast-slow proxy types. For example, in the snippet below, the `xpd.Series` type is an instance `_FastSlowProxyMeta`. Therefore we can access the property `_fsproxy_fast` defined in the metaclass.

Reviewer 2: ```suggestion
`cudf.pandas` uses a [metaclass](https://docs.python.org/3/glossary.html#term-metaclass) called (`_FastSlowProxyMeta`) to dynamically find class attributes and classmethods of fast-slow proxy types. For example, in the snippet below, the `xpd.Series` type is an instance `_FastSlowProxyMeta`. Therefore we can access the property `_fsproxy_fast` defined in the metaclass.
```

```python
import cudf.pandas
cudf.pandas.install()
import pandas as xpd

print(xpd.Series._fsproxy_fast) # output is cudf.core.series.Series
```

### Caching

> ### Pickling and Unpickling

Author: Do you think I should document how these two work in `cudf.pandas`?

Reviewer 2: Yes, but I would be OK merging this PR without that. Pickling in cudf.pandas is a pretty fraught topic that is worth a separate PR where we also iron out exactly what the current limitations are.
